core/tests/unit/kmnkbd/test_actions_normalize.cpp

```diff
@@ -0,0 +1,299 @@
+    /* cached context post transform: */ u"abcA\u0300" u"e\u0316\u0301" u"\u0073\u0323\u0307"  u"\u0041\u030a"     u"\U000114B9\U000114B0",
+    /* action del, output: */            0, U"A\u0300" U"e\u0316\u0301" U"\u0073\u0323\u0307"  U"\u0041\u030a"     U"\U000114B9\U000114B0",
+    // ---- results ----
+    /* action del, output: */            0, U"À"       U"é̖"             U"\u1e69"              U"\u00c5"           U"\U000114BC",
+    /* app_context: */                   u"abcÀé̖\u1e69\u00c5\U000114BC"
+  );
+
+  // Interaction with input context when not on normalization boundary
+
+  test(
+    "Backtrack one character to combine as NFC",
+    /* app context pre transform: */     u"XYZA",
+    /* cached context post transform: */ u"XYZA\u0300abc",
+    /* action del, output: */            0, U"\u0300abc",
+    // ---- results ----
+    /* action del, output: */            1, U"Àabc",
+    /* app_context: */                   u"XYZÀabc"
+  );
+
+  test(
+    "Backtrack e+comb circ (2 chars) to combine as NFC",
+    /* app context pre transform: */     u"abce\u0302",
+    /* cached context post transform: */ u"abce\u0323\u0302",
+    /* action del, output: */            1, U"\u0323\u0302",
+    // ---- results ----
+    /* action del, output: */            2, U"ệ",
+    /* app_context: */                   u"abcệ"
+  );
+
+  test(
+    "Two backspaces for NFD converts into one char in NFC (ê) and recombine",
+    /* app context pre transform: */     u"abcê",
+    /* cached context post transform: */ u"abce\u0323\u0302",
+    /* action del, output: */            2, U"e\u0323\u0302", // NFD input;  delete 2: e and \u0302
+    // ---- results ----
+    /* action del, output: */            1, U"ệ",             // NFC output; delete 1: ê
+    /* app_context: */                   u"abcệ"
+  );
+
+  // a\u0300 should not be normalized because it is not otherwise impacted by
+  // the action.
+  test(
+    "Avoid editing too far back in context when finding normalization boundary",
+    /* app context pre transform: */     u"a\u0300bcê",
+    /* cached context post transform: */ u"a\u0300bce\u0323\u0302",
+    /* action del, output: */            2, U"e\u0323\u0302", // NFD input;  delete 2: e and \u0302
+    // ---- results ----
+    /* action del, output: */            1, U"ệ",              // NFC output; delete 1: ê
+    /* app_context: */                   u"a\u0300bcệ"
+  );
+
+  // If we don't reach a normalization boundary, we still should continue to work
+  test(
+    "Normalizable letters at start of context",
+    /* app context pre transform: */     u"\u0300",
+    /* cached context post transform: */ u"\u0300\u0323\u0302",
+    /* action del, output: */            0, U"\u0323\u0302",       // NFD input;
+    // ---- results ----
+    /* action del, output: */            1, U"\u0323\u0300\u0302",  // NFC output is still decomposed because there is no base
+    /* app_context: */                   u"\u0323\u0300\u0302"
```

Comment:
I would have expected the output and app_context to be `u"\u0300\u0323\u0302"`. Or did the normalization change the order? But then I would expect the cached context to have the same order.

Reply:
Yes, the normalization should put `\u0323` first. Again, relates to #10422 discussion, my misunderstanding on what we could expect LDML keyboard processor to generate. Updated this test to match input assumption that cached_context is always NFD.